### ISSUE TITLE
Korean is broken in the game title.

### ISSUE THREAD

**Issue by kbj1213** [icza/screp#46]
Hi, I'm using screp.

But I have a problem.

When the game's title and map's name is in Korean, the string is broken.

I'm guessing it's an encoding issue.

Could it be revised in the next version?

**Comment by icza** [comment]
Sure, I can check it. Please send example replays.

**Comment by kbj1213** [comment]
Here it is!
https://repmastered.app/game/ZIgQt8StxDLueYfqv8-C92w2mwEWaOXRRPkCQuSUFDI

**Comment by icza** [comment]
The title of that game is decoded to: 3:3 鍮⑤Т 諛곗냽 �쒕�

What should it be? What is the correct title?

**Comment by kbj1213** [comment]
I don't know exactly but it starts like "3:3 빨무 .... "

**Comment by icza** [comment]
I checked this replay. Yes, the problem is a character encoding issue.

screp has a simple, builtin encoding detector. Basically if the raw data is a valid UTF-8 encoded text, then UTF-8 decoding is used to obtain the text. If it is not, EUC-KR (also known as Code Page 949) is also tried, and if it succeeds (no error decoding raw data using EUC-KR), then this will be the result.

In this replay, the raw title bytes are:

    [51 58 51 32 235 185 168 235 172 180 32 235 176 176 236 134 141 32 235 147 156 235 163 0 0 0 0 0]

It's a 0-terminated string, so trailing zeros are cut off:

    [51 58 51 32 235 185 168 235 172 180 32 235 176 176 236 134 141 32 235 147 156 235 163]

The thing is that this is _supposed to be_ UTF-8, because decoding it as UTF-8 results in:

    3:3 빨무 배속 드�

(Note the unicode replacement character (the question mark) at the end.)
BUT this is INVALID UTF-8, so screp falls back to EUC-KR, and decoding it as such results in:

    3:3 鍮⑤Т 諛곗냽 �쒕�

To sum it up: the raw data is supposed to be UTF-8, but the data is invalid. So screp falls back to a wrong encoding.

**Comment by icza** [comment]
Note that there are many Korean examples where decoding works. It would also work here using UTF-8 if the raw data wouldn't be invalid UTF-8.

**Comment by icza** [comment]
For example check out this game: https://repmastered.app/game/HXMAHL9rvAYiyTAgRxe2RIl87upz5XKq-U2wRGDdYmM

The game title is Korean, there's a player with Korean name (a computer), and there is also Korean chat, all properly decoded.

**Comment by armoha** [comment]
I think we could make an exception, to not fallback to CP949 (EUC-KR) if only last character of game title is invalid UTF-8.

**Comment by icza** [comment]
I'd rather like to understand the cause first before adding any exceptions.

If this is supposed to be UTF-8 encoded, why is it invalid UTF-8? Is it a bug in StarCraft or something else? Generating UTF-8 encoded data isn't something rare or complex, and StarCraft does it everywhere else properly. Does this happen if certain characters used in the title?

**Comment by armoha** [comment]
Yeah, I also want to know exactly what is happening.

> If this is supposed to be UTF-8 encoded, why is it invalid UTF-8?

First I suspected because game tittle is too long. Game title can be up to 31 characters long in battle.net (not byte limit), which is much larger than replay header stores.

So I tried [very](https://repmastered.app/game/eo560Bg93b-q25oXT9541D9x5rbNgeptxqe0AJGgoOY) [long](https://repmastered.app/game/YgUi73HScl4N_RLw41QghTaReZ87OleDusfjH8UdG4Q) [game title](https://repmastered.app/game/LIp86sVnYcEGlEv5qsppkeICIHAJyF7QVts-PXYwxKY) but couldn't reproduce the problem.

> Does this happen if certain characters used in the title?

No idea, I downloaded and tested the replay but StarCraft: Remastered does not show game title of replay so can't guess what original game titile was.

**Comment by icza** [comment]
If there's not enough space for the title in the header, yes, the title must be truncated, which StarCraft may do so in the "middle" of an UTF-8 sequence (I haven't checked).

But there are 28 bytes reserved for map title in the header, and the "truncated" title doesn't even use all that:

    [51 58 51 32 235 185 168 235 172 180 32 235 176 176 236 134 141 32 235 147 156 235 163 0 0 0 0 0]

This uses only 23 bytes, and there are 5 zeroes remaining. So it makes no sense to cut the title data in the middle of a multi-byte UTF-8 sequence...

Note that if there was an additional, non-zero byte anything from the range [128..191], it would be a valid UTF-8 encoded sequence. You can check them here: https://go.dev/play/p/-GigvCAEdV1

I don't speek Korean, but many of them seems to be a sensible title: [Google translations](https://translate.google.com/?sl=auto&tl=en&text=128%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%80%0A129%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%81%0A130%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%82%0A131%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%83%0A132%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%84%0A133%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%85%0A134%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%86%0A135%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%87%0A136%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%88%0A137%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%89%0A138%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8A%0A139%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8B%0A140%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8C%0A141%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8D%0A142%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8E%0A143%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%8F%0A144%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%90%0A145%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%91%0A146%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%92%0A147%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%93%0A148%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%94%0A149%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%95%0A150%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%96%0A151%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%97%0A152%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%98%0A153%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%99%0A154%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9A%0A155%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9B%0A156%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9C%0A157%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9D%0A158%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9E%0A159%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%9F%0A160%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A0%0A161%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A1%0A162%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A2%0A163%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A3%0A164%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A4%0A165%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A5%0A166%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A6%0A167%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A7%0A168%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A8%0A169%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%A9%0A170%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AA%0A171%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AB%0A172%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AC%0A173%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AD%0A174%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AE%0A175%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%AF%0A176%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B0%0A177%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B1%0A178%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B2%0A179%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B3%0A180%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B4%0A181%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B5%0A182%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B6%0A183%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B7%0A184%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B8%0A185%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%B9%0A186%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BA%0A187%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BB%0A188%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BC%0A189%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BD%0A190%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BE%0A191%203%3A3%20%EB%B9%A8%EB%AC%B4%20%EB%B0%B0%EC%86%8D%20%EB%93%9C%EB%A3%BF&op=translate)

Some examples:

```
130 3:3 빨무 배속 드룂    fast speed drone
131 3:3 빨무 배속 드룃    fast speed drone
141 3:3 빨무 배속 드룍    speed drop
```

What happens if you use some of these titles?

**Comment by icza** [comment]
OK, I think I have a solution here. SC:R always uses UTF-8. So this can be decoded using forced UTF-8 since this was played using SC:R. Will update the parser.

**Comment by icza** [comment]
Released screp v1.11.3, forcing UTF-8 when reading game title: https://github.com/icza/screp/releases/tag/v1.11.3
